File: src/helpers/HashHelper.java
/**
 * Copyright 2013 ShkSchneider
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package me.shkschneider.skeleton.helpers;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

public abstract class HashHelper {

    public static final String MD5 = "MD5";
    public static final String SHA = "SHA";

	private static String hash(final String string, final String algorithm, final Integer length) {
		try {
			MessageDigest messageDigest = MessageDigest.getInstance(algorithm);
			StringBuilder stringBuilder = new StringBuilder();

			messageDigest.reset();
			messageDigest.update(string.getBytes());
			byte digest[] = messageDigest.digest();
			stringBuilder.setLength(0);
			for (int i = 0; i < digest.length; i++) {
				final int b = digest[i] & 255;
				if (b < length) {
					stringBuilder.append('0');
				}
				stringBuilder.append(Integer.toHexString(b));
			}
			return stringBuilder.toString();
		}
		catch (NoSuchAlgorithmException e) {
			LogHelper.e("NoSuchAlgorithmException: " + e.getMessage());
		}
		return null;
	}

	public static String md5(final String string) {
		return hash(string, HashHelper.MD5, 32);
	}

	public static String sha(final String string) {
		return hash(string, HashHelper.SHA, 40);
	}

}

File: android/src/main/kotlin/me/shkschneider/skeleton/android/security/HashHelper.kt
package me.shkschneider.skeleton.android.security

import me.shkschneider.skeleton.android.log.Logger
import java.math.BigInteger
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException

object HashHelper {

    private const val MD5 = "MD5"
    private const val SHA1 = "SHA-1"
    private const val SHA256 = "SHA-256"
    private const val SHA512 = "SHA-512"

    private fun hash(algorithm: String, string: String): String? {
        try {
            with(MessageDigest.getInstance(algorithm).digest(string.toByteArray())) {
                return BigInteger(1, this)
                        .toString(16).padStart(32, '0')
            }
        } catch (e: NoSuchAlgorithmException) {
            Logger.wtf(e)
            return null
        }
    }

    @Deprecated("Unsafe.")
    fun md5(string: String): String? {
        return hash(MD5, string)
    }

    @Deprecated("Unsafe.")
    fun sha1(string: String): String? {
        return hash(SHA1, string)
    }

    fun sha256(string: String): String? {
        return hash(SHA256, string)
    }

    fun sha512(string: String): String? {
        return hash(SHA512, string)
    }

}
